docs/user-guide/memory.md

```diff
@@ -0,0 +1,48 @@
+# Memory
+
+Cubed aims to tightly manage the amount of memory that each task uses, ensuring that it never exceeds a configured maximum.
+
+The various memory settings and values are illustrated in the following diagaram, and are discussed further below.
+
+![Memory](../images/memory.svg)
+
+## Allowed memory
+
+You should set ``allowed_mem`` to the maximum amount of memory that is available to the Cubed runtime. When running locally this should be no more than the amount of memory you have available on your machine. For cloud services it should be the amount of memory that the container runtime is configured to use.
+
+In this example we set the allowed memory to 2GB:
+
+```python
+import cubed
+
+spec = cubed.Spec(allowed_mem="2GB")
+```
+
+## Projected memory
+
+The output of a Cubed computation is a chunked Zarr array (or set of arrays). Tasks are run by workers to operate on Zarr chunks, and the memory needed for each task is a function of chunk size, dtype, and the nature of the operation, which can be computed while building the plan.
+
+While it is not possible in general to compute the precise amount of memory that will be used (since it is not known ahead of time how well a Zarr chunk will compress, for example), it is possible to put a conservative upper bound on the memory usage. This upper bound is the `projected_mem` for a task, and it is calculated automatically by Cubed.
+
+If the `projected_mem` calculated by Cubed is greater than the value of `allowed_mem` that you have set, then an exception is raised during the planning phase. This check means that you can have high confidence that the operation will run within its memory budget.
+
+## Reserved memory
+
+There is another memory setting that can help Cubed manage memory properly: `reserved_mem`. This is the amount of memory reserved on a worker for non-data use - it's whatever is needed by the Python process for running a task.
+
+Cubed provides a utility function, {py:func}`measure_reserved_mem <cubed.measure_reserved_mem>`, for measuring it for the runtime you are using. You can run this function first, then use the value it reports as the basis for setting `reserved_mem` for your computation. Cubed will use `reserved_mem` as a baseline when calculating `projected_mem`, in order to more accurately estimate the upper bound on memory usage, making it more likely that your computation doesn't exceed memory bounds.
+
+Some executors have the ability to measure memory as a task is running. They can report the `peak_measured_mem`, which is the highest actual memory usage for a given task during its execution. This information is useful for analysing how well the computation ran, and if there is room for improvement (in Cubed's algorithms for calculating `projected_mem`).
+
+## Chunk sizes
+
+As a rule of thumb `allowed_mem` should be at least ten times as large as your Zarr chunk sizes.
```

Comment:
Do your cubed chunk sizes have to be the same as your Zarr chunk sizes? And if not then what size chunk is likely to work well? Dask says 100MB as a general rule of thumb, but it might be different for cubed...

Reply:
Cubed uses Zarr for all its array representations - so the chunk sizes are the same. If you started with a Zarr array with very large chunks then you might want to rechunk them to something smaller or better shaped before doing a reduction.

I think 100MB is probably a good rule of thumb - fast enough to retrieve/process/store, and also ~20 fit into typical allowed mem (2GB), which is good for reductions (as mentioned here).

Reply:
> I think 100MB is probably a good rule of thumb

Okay great, that's definitely worth adding (and also that 2GB is a reasonable `allowed_mem`).

> Cubed uses Zarr for all its array representations - so the chunk sizes are the same.

Right, I get the `cubed.Array` objects are backed by Zarr arrays of the exact same chunk sizes, but what about when you first call `cubed.from_array`? With dask it is possible to choose different chunks than those on disk, and sometimes there are good reasons to (e.g. combine multiple tiny chunks into 1 big one). Is that possible with cubed?

Reply:
Yes, you can pass `chunks` to `from_array` and it will rechunk if necessary. See https://github.com/tomwhite/cubed/blob/7ccd99158bb17d23cd9ff8bc90841aff11c76685/cubed/core/ops.py#L33-L72. 

BTW I just noticed that this path doesn't have an Xarray fast path in the case that Xarray is backed by Zarr arrays. Is this a path that cubed-xarray uses?